### ISSUE TITLE
[Console] Command::addArgument accepts mixed $default, passed to InputArgument constructor not accepting mixed

### ISSUE THREAD

**Issue by Ahummeling** [symfony/symfony#45110]
### Symfony version(s) affected

6.0.0 6.1.x-dev

### Description

Consider my method: `Command::configure`:
```php
    protected function configure(): void
    {
        $this->setName('app:admin:create')
            ->setDescription('Creates a new admin user')
            ->addArgument('username', InputArgument::REQUIRED, 'Sets the username of the user to create')
            ->addArgument('password', InputArgument::OPTIONAL, 'Sets the initial password for the user', fn() => $this->passwordGenerator->generatePassword());
    }
```

This leads to the following error:
```shell
Symfony\Component\Console\Input\InputArgument::__construct(): Argument #4 ($default) must be of type array|string|int|float|bool|null, Closure given
```

### How to reproduce

See code in description
Running `bin/console --version` was enough to trigger the type error. 

Take a skeleton 6.0 symfony project, include the following snippet, and try running `bin/console cache:clear -vvv` for full details.
```php
<?php
namespace App\Command;

class PasswordGenerator 
{
    public function generatePassword(): string
    {
        return 'password';
    }
}

use Symfony\Component\Console\Command\Command;
class AppCommand extends Command
    private PasswordGenerator $passwordGenerator;
    protected function configure(): void
    {
        $this->passwordGenerator = new PasswordGenerator();
        $this->setName('app:admin:create')
            ->setDescription('Creates a new admin user')
            ->addArgument('username', InputArgument::REQUIRED, 'Sets the username of the user to create')
            ->addArgument('password', InputArgument::OPTIONAL, 'Sets the initial password for the user', fn() => $this->passwordGenerator->generatePassword());
    }
}
```

### Possible Solution

I'm actually not sure about the best way to approach this, a few ideas popped into my mind:
 - Widen the parameter type in `InputArgument::__construct` to accept mixed (no BC break)
 - Narrow the parameter type in `Command::addArgument` to also only accept `string|bool|int|float|array|null` (introduces BC break)
 - Narrow the parameter annotation as in the solution above, but don't change the actual function signature and then trigger a deprecation warning.
 - Validate the passed `$defaultValue` for the correct type and throw an `InvalidArgumentException` in case it's an `object|resource|closure`. This keeps BC in the function signatures. But it doesn't really accomplish much.

### Additional Context

Would be happy to work out a solution and submit a PR if this is indeed something to be fixed.
Also, I am well aware I can transform my closure into a callable array and have it work, but the point of this issue is to point out the inconsistency in argument types.

**Comment by kbond** [comment]
I am seeing some ambiguity in the return types for arguments/options. For instance, `InputInterface::getArguments(): array<string|bool|int|float|array|null>` but `InputInterface::getArgument(): mixed`. This long chain of union types feels like @nicolas-grekas went out of his way to exclude some types in https://github.com/symfony/symfony/pull/41991 - I'm not sure if this was by design.

**Comment by wouterj** [comment]
Hi!

I'm not sure we should accept non-scalar values as default value. It's never possible to pass a function or object via command line arguments/options. This is why the types are limited to all scalar values (or a list of scalar values). This isn't regarded as a BC break, as the contract defined by the PHPdoc has been `string|string[]|null` since 2.8 (https://github.com/symfony/symfony/pull/28647). While it was possible to pass a function, it was not officially supported.

Cases like this are often covered the same way as you would do this with PHP arguments: using `null` as a default value and doing `$option = $option ?? $this->passwordGenerator->generatePassword()` in `Command::execute()`.

If you think it's nice to allow this type of "lazily defined default values", this can probably considered as a new feature for 6.1.

**Comment by Ahummeling** [comment]
I think I completely agree with you in the sense that a default value should be of a type that cannot be passed via the command line. And with that in mind, I think the accepted type of the `$defaultValue` parameter should probably change from `mixed` to `string|bool|int|float|array|null`. 

But if I were to submit a PR like [this](https://github.com/symfony/symfony/pull/45115), I think it would be a BC break, because any class extending from `Command` overriding the `addArgument` or `addOption` method, will now fail on autoloading due to a mismatch in parameter types, right? 

I noticed that the type hint of the phpdoc in the console was changed in the effort to updating to typed parameters in this [commit](https://github.com/symfony/symfony/commit/66b686503792392d5b472f3b1402d23612902d72). Not sure if this implies the contract changed with it?

I am not sure if the added complexity of these lazily defined / evaluated default values is worth it, although I suppose there's an intuitive feel to providing a closure for resolving the default value, should none be provided. Would opening an issue as [RFC] be the appropriate place to discuss whether such a feature has a place in the code base? I would be more than happy to provide a PoC implementation alongside this issue, although I'm not entirely sure I'll get it right the first time.
